src/docker.cc

```diff
@@ -50,77 +127,56 @@ std::vector<MetadataUpdater::ResourceMetadata>
-        if (config_.VerboseLogging()) {
-          LOG(INFO) << "Inspect response: " << body(inspect_response);
-        }
-        json::value parsed_metadata =
-            json::Parser::FromString(body(inspect_response));
-        if (config_.VerboseLogging()) {
-          LOG(INFO) << "Parsed metadata: " << *parsed_metadata;
-        }
-
-        const MonitoredResource resource("docker_container", {
-          {"location", zone},
-          {"container_id", id},
-        });
-
-        const json::Object* container_desc = parsed_metadata->As<json::Object>();
-        const std::string name = container_desc->Get<json::String>("Name");
-
-        const std::string created_str =
-            container_desc->Get<json::String>("Created");
-        Timestamp created_at = rfc3339::FromString(created_str);
-
-        const json::Object* state = container_desc->Get<json::Object>("State");
-        bool is_deleted = state->Get<json::Boolean>("Dead");
-
-        const std::string resource_id =
-            std::string("container") + config_.MetadataApiResourceTypeSeparator() + id;
-        // The container name reported by Docker will always have a leading '/'.
-        const std::string resource_name =
-            std::string("container") + config_.MetadataApiResourceTypeSeparator() + name.substr(1);
-        result.emplace_back(std::vector<std::string>{resource_id, resource_name},
-                            resource,
-                            MetadataAgent::Metadata(config_.DockerApiVersion(),
-                                                    is_deleted, created_at, collected_at,
-                                                    std::move(parsed_metadata)));
+        const json::Object* container = raw_container->As<json::Object>();
+        result.emplace_back(GetContainerMetadata(container, collected_at));
       } catch (const json::Exception& e) {
         LOG(ERROR) << e.what();
         continue;
       }
     }
   } catch (const json::Exception& e) {
     LOG(ERROR) << e.what();
-  } catch (const boost::system::system_error& e) {
-    LOG(ERROR) << "Failed to communicate with " << docker_endpoint << ": " << e.what();
+  } catch (const QueryException& e) {
+    // Already logged.
   }
   return result;
 }
 
+json::value DockerReader::QueryDocker(const std::string& path) const
+    throw(QueryException, json::Exception) {
+  const std::string endpoint(config_.DockerEndpointHost() +
+                             "v" + config_.DockerApiVersion() +
+                             path);
+  http::local_client client;
+  http::local_client::request request(endpoint);
+  if (config_.VerboseLogging()) {
+    LOG(INFO) << "QueryDocker: Contacting " << endpoint;
+  }
+  try {
+    http::local_client::response response = client.get(request);
+#ifdef VERBOSE
+    LOG(DEBUG) << "QueryDocker: Response: " << body(response);
+#endif
+    return json::Parser::FromString(body(response));
```

Comment:
This is the area I'm referring to for non 200 responses. This is likely work for the future, but we should be checking for the response status code, instead of just assuming we're returning the right response based on ANY response at all.